internal/pagefile_test.go

```diff
@@ -3,75 +3,262 @@ package internal_test
+	content := createPageFileContent(1, pagedFileBlockCount)
+	mockFile := NewMockReaderAtWriterAt(content)
+
+	err := internal.WritePagesFromIncrement(testIncrement.NewReader(), mockFile, false)
+
+	assert.NoError(t, err)
+	// check that no bytes were written to the mock file
+	assert.Equal(t, 0, mockFile.bytesWritten)
+}
+
+// In this test series we test that
+// all increment blocks are being written if the
+// local file is empty (all blocks are missing)
+func TestWritingIncrementToEmptyFile(t *testing.T) {
+	postgresWritePagesTestEmptyFile(regularTestIncrement, t)
+}
+
+func TestWritingIncrementToEmptyFileSmallLSN(t *testing.T) {
+	postgresWritePagesTestEmptyFile(smallLsnTestIncrement, t)
+}
+
+func TestWritingIncrementToEmptyFileBigLSN(t *testing.T) {
+	postgresWritePagesTestEmptyFile(bigLsnTestIncrement, t)
+}
+
+func postgresWritePagesTestEmptyFile(testIncrement *TestIncrement, t *testing.T) {
+	content := make([]byte, internal.DatabasePageSize*pagedFileBlockCount)
+	mockFile := NewMockReaderAtWriterAt(content)
+	err := internal.WritePagesFromIncrement(testIncrement.NewReader(), mockFile, false)
+	assert.NoError(t, err)
+	assert.Equal(t, testIncrement.fileSize, uint64(len(mockFile.content)))
+
+	sourceFile, _ := os.Open(pagedFileName)
+	defer utility.LoggedClose(sourceFile, "")
+	checkIfAllIncrementBlocksExist(mockFile, sourceFile, testIncrement.diffBlockCount,t)
+}
+
+func TestRestoringPagesToCompletedFile(t *testing.T) {
+	pagedFile, _ := os.Open(pagedFileName)
+	fileReader := io.Reader(pagedFile)
+	defer utility.LoggedClose(pagedFile, "")
+	content := createPageFileContent(1, pagedFileBlockCount)
+	mockFile := NewMockReaderAtWriterAt(content)
+
+	err := internal.RestoreMissingPages(fileReader, mockFile)
+
+	assert.NoError(t, err)
+	// check that no bytes were written to the mock file
+	assert.Equal(t, 0, mockFile.bytesWritten)
+}
+
+func TestRestoringPagesToEmptyFile(t *testing.T) {
+	pagedFile, _ := os.Open(pagedFileName)
+	fileReader := io.Reader(pagedFile)
+	defer utility.LoggedClose(pagedFile, "")
+	content := make([]byte, internal.DatabasePageSize*pagedFileBlockCount)
+	mockFile := NewMockReaderAtWriterAt(content)
+
+	err := internal.RestoreMissingPages(fileReader, mockFile)
+
+	assert.NoError(t, err)
+	mockFileReader := bytes.NewReader(mockFile.content)
+	pagedFile.Seek(0,0)
+	compareResult := deepCompareReaders(pagedFile, mockFileReader)
+	assert.Truef(t, compareResult, "Increment could not restore file")
+}
+
+func checkIfAllIncrementBlocksExist(mockFile *MockReadWriterAt, sourceFile io.ReaderAt,
+	diffBlockCount uint32, t *testing.T) {
+	emptyPage := make([]byte, internal.DatabasePageSize)
+	emptyBlockCount := uint32(0)
+	dataBlockCount := uint32(0)
+
+	for index, data := range mockFile.getBlocks() {
+		readBytes := make([]byte, internal.DatabasePageSize)
+		sourceFile.ReadAt(readBytes, index*internal.DatabasePageSize)
+		if bytes.Equal(emptyPage, data) {
+			emptyBlockCount += 1
```

Comment:
You don't check that dataBlockCount + emptyBlockCount is equal to totalBlockCount

Reply:
Prior to this check I verify that resulting file size is equal to the correct file size. I redesigned this function a bit in ed861e47a77988113cb0d21f7b58b6c483779efe. 

I check that:

1. All of the increment blocks exists in file
2. Each increment block is in the correct place
3. No extra data except the empty blocks exist in file
4. File size match the required file size

If we check all of that, it is automatically true that dataBlockCount + emptyBlockCount is equal to totalBlockCount